external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpout.java

```diff
@@ -383,23 +403,24 @@ private void commitOffsetsForAckedTuples() {
     @Override
     public void ack(Object messageId) {
         final KafkaSpoutMessageId msgId = (KafkaSpoutMessageId) messageId;
-        if(!emitted.contains(msgId)) {
-            LOG.debug("Received ack for tuple this spout is no longer tracking. Partitions may have been reassigned. Ignoring message [{}]", msgId);
-            return;
-        }
-        
-        if (!consumerAutoCommitMode) {  // Only need to keep track of acked tuples if commits are not done automatically
-            acked.get(msgId.getTopicPartition()).add(msgId);
+        if (!emitted.contains(msgId)) {
+            LOG.debug("Received ack for message [{}], which is associated with either a null tuple that was never emitted, " +
```

Comment:
I mean that the debug log here is now getting written both when the spout receives an ack for a tuple from a partition that has been reassigned since it was emitted, and when the RecordTranslator emits a null tuple. If you add msgId to emitted in L357 before calling ack(), we don't have to change the log here to also mention null tuples.

I think logging something when null tuples are being dropped might fit better in RecordTranslator.apply (or in L357) than here, because then the log can be clear about what is happening.